File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    # 간선 길이가 동일 -> BFS 해도 되겠다!
    # 인접 리스트부터 구하기!
    adj = [[] * (n+1) for _ in range(n+1)]
    visited = [False] * (n+1)
    for n1, n2 in edge:
        adj[n1].append(n2)
        adj[n2].append(n1)
    
    # BFS - 1번 노드 시작, (노드, dist) 넣어서 maxDist를 갱신하면 될 듯!
    maxDist = 0
    maxNodeCnt = 0  
    q = deque([(1,0)])
    visited[1] = True
    
    while q:
        cn, dist = q.popleft()

        if maxDist < dist:
            maxDist = dist
            maxNodeCnt = 1
        elif maxDist == dist:
            maxNodeCnt += 1
        
        for n in adj[cn]:
            if not visited[n]:
                q.append((n, dist + 1))
                visited[n] = True
        
    return maxNodeCnt

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int n, vector<vector<int>> edge) {
    // 인접 리스트 만들기
    vector<vector<int> > adj(n+1);
    
    for (auto nodes : edge) {
        auto v1 = nodes[0];
        auto v2 = nodes[1];
        adj[v1].push_back(v2);
        adj[v2].push_back(v1);
    }
    
    // bfs 돌리기 - (node, dist) -> 가장 먼 길이, 가장 먼 노드 개수
    queue<pair<int, int> > q;
    vector<bool> visited(n+1, false);
    
    q.push({1, 0});
    visited[1] = true;
    
    int maxDist = 0;
    int maxNodeCnt = 0;
    
    while (!q.empty()) {
        auto [cn, dist] = q.front();
        q.pop();
        
        // 가장 먼 길이 및 노드 개수 갱신
        if (maxDist < dist) {
            maxDist = dist;
            maxNodeCnt = 1;
        } else if (maxDist == dist) {
            maxNodeCnt ++;
        }
        
        for (auto n : adj[cn]) {
            if (!visited[n]) {
                visited[n] = true;
                q.push({n, dist+1});
            }
        }
    }
    
    return maxNodeCnt;
}